persistence/docs/SAVEPOINTS_ROLLBACKS.md

```diff
@@ -0,0 +1,146 @@
+
+This means that technically each one of these trees is a separate data store.
+
+### Data Consistency (the "C" in CAP Theorem)
+
+We cannot really make the assumption, especially in this day and age, but in the simplest case, we could very much have a monolytic setup where the node is running on the same machine as the `PostgresSQL` database and the `BadgerDB` key-value stores but that would not change the fact that we are dealing with a distributed system since each one of these components is a separate process that could fail independently.
+
+Imagine a scenario in which the state is committed to the `PostgresSQL` database but one/some of the `BadgerDB` key-value stores fails due to storage issues. What would happen next? That's **non-deterministic**.
+
+Even a single node is a small distributed system because it has multiple separate components that are communicating with each other (on the same machine or via unreliable networks).
+
+Since a node is part of a distributed network of nodes that are all trying to reach consensus on the "state of the world", we have to make sure that the data that we are storing is consistent internally in the first place.
+
+In the event of a failure at any level during this process, we cannot commit state non-atomically. That would make the node inconsistent and put it in a non-deterministic state that is not recoverable unless we have a way to rollback to a previous, clean, state with all the implications that this would have on the network. (social coordination comes to mind)
+
+We either **succeed** across **all** data stores or we **fail** and we have to be able to **recover** from that failure like if **nothing has happened**.
+
+
+The following diagram illustrates the high-level flow:
+
+```mermaid
+flowchart TD
+    NewBlock[A Block was just committed\nor we are at Height 0] -->CreateSavepoint(Create SavePoint)
+    CreateSavepoint --> UpdatingState(We are creating\na new block.\nUpdating the State)
+    UpdatingState --> ShouldCommit
+    ShouldCommit{Should we commit?} --> |Yes| ReceivedOKFromDatastores{Have all\nthe datastores\ncommitted\nsuccessfully?}
+    ShouldCommit --> |No| NewBlock
+    ReceivedOKFromDatastores -->|Yes| Committed[Committed successfully] --> |Repeat with\nthe new Block| NewBlock
+    ReceivedOKFromDatastores -->|No, we have at least one failure| Rollback[Rollback] --> |Repeat with\nthe rolled back Block | NewBlock
+
+```
+
+## Definitions
+
+### Savepoints
+
+A savepoint (also referred as checkpoints/snapshots depending on the implementation) is either the beginning of database transaction (or distributed transaction) or some sort of artifact that was created right after a successful commit happened that allows to recreate a perfect copy of the state at the time it was created.
```

Comment:
TL;DR: yes

If the process that writes the files in the datadirs was stopped gracefully and there are no open handles on these files, a timestamped copy of these datadirs is what I would call `snapshot`.

Does this answer your question?

Reply:
It does answer my question, but I believe we are conflating `snapshot` and `savepoint`.

My original intention was:
* `snapshot` - A durable, shareable copy of the data directory.
* `savepoint` - an in-memory state that can be rolled back, updated or committed. It may or may not directly correspond to a `snapshot` that was stored on disk.